blockchain/blockchain.go

```diff
@@ -411,6 +419,10 @@ func (chain *Blockchain) applyGlobalParams(appState *appstate.AppState, block *t
 	if flags.HasFlag(types.ValidationFinished) {
 		appState.State.SetValidationPeriod(state.NonePeriod)
 	}
+	if block.Height()-appState.State.LastSnapshot() >= chain.config.Consensus.SnapshotRange && appState.State.ValidationPeriod() == state.NonePeriod &&
```

Comment:
Code is very similar to calculateFlags. Maybe here we can check flag in block only?